### ISSUE TITLE
Refactor for ECMASCRIPT 6 proxying, async/await, etc

### ISSUE THREAD

**Issue by formix** [formix/infernal-engine#41]
Within a rule, the this keyword shall be set to a proxy to be able to build the fact context using the following syntax:

this.$ refers to the root context.
this._ refers to the parent context.
this._ctx returns the current context name. Ex.: "/engine/cylinders" for a rule named "/engine/cylinders/validateSize".
this._name returns the current rule name. Ex: "/engine/cylinders/validateSize".
this.whatever gets or sets the value of the whatever fact in the current context.
this.child.grandchild gets or sets the value of the grandchild fact from the current context's child.
Example 1: if a rule is in the "/engine/cylinders" context, doing this._.torque allows to get or set the "/engine/torque" fact.

Example 2: from the same rule in the "/engine/cylinders" context, doing this.$.engine.torque would do the same as above.

**Comment by formix** [comment]
Within a rule, the `this` keyword shall be set to a proxy to be able to build the fact context using the following syntax:

1. `this.$` refers to the root context.
2. `this._` refers to the parent context.
3. `this._ctx` returns the current context name
4. `this.whatever` gets or sets the value of the `whatever` fact in the current context.
5. `this.child.grandchild" gets or sets the value of the grandchild fact from the current context's child.

Example 1: if a rule is in the "/engine/cylinders" context, doing `this._.torque` allows to get or set the "/engine/torque" fact.

Example 2: from the same rule in the "/engine/cylinders" context, doing `this.$.engine.torque` would do the same as above.

**Comment by formix** [comment]
Totally screed up the "this" usage in favor of idem potent rules with a parameter annotation syntax to reference rules and returning a javascript object ex: `{"fact/path1": "value1", "fact/path2": 2}`.

by default parameter names refer to facts of the same name in the same context or leverage an annotation comment to know the actual rule path. Ex:

```javascript
function rule(localFact1, /*@ /remote/fact/somewhere */ anotherFact) {
  // ...
}
```

**Comment by formix** [comment]
Fixed by PR #46
